proposals/0000-linear-types.rst

```diff
@@ -0,0 +1,2290 @@
+
+That is, in the above example, ``R`` has type
+
+::
+
+  R :: A ->. B ->. R
+
+In general, in
+
+::
+
+  data R where
+    R :: { f1 ::(π) A, f2 ::(ρ) B } -> R
+
+We have
+
+::
+
+  R :: A :π-> B :ρ-> R
+
+With absence of annotation interpreted as annotating with ``'One``.
+
+
+Inference
+---------
+
+.. _Inference:
+
+Because of backwards compatibility, we initially chose the following
+strategy: when the type of a function is not constrained by given
+constraints, we conservatively assume it to have multiplicity ω.
+
+Experience shows that this sometimes yield very confusing error messages
+where perfectly valid code is rejected:
+
+::
+
+  type family L x
+  type instance L Int = A ->. A
+
+  f :: L x -> x
+
+  u :: Int
+  u = f (\x -> x)
+
+While the identity function is indeed linear, because the resolution
+of the type family (``L Int ~ Int``) is delayed in GHC, ``\x -> x`` is
+considered to have no given type, and is inferred to have a non-linear
+type, and is refused by the type-checker.
+
+We therefore need a more refined strategy, to avoid surprising
+behaviour like the above. We do not expect it to be too hard to
+implement a better strategy, but we don't have a specification yet.
+
+A more profound difficulty exists for inference: for explicit ``let``
+bindings and ``case`` expressions (*i.e.* which are not generated from
+the desugaring of an equation but are written as ``let``, ``where``,
+or ``case`` in the surface syntax), we want to infer the multiplicity
+annotation. The process for this is not yet defined (see `Unresolved
+questions`_ below for a more precise description of this issue).
+
+Non-termination, exceptions & catch
+-----------------------------------
+
+.. _Exceptions:
+
+In the presence of non-termination or exceptions, linear functions may
+fail to fully consume their argument. We can think of it as: the
+consumption of the result of the function was never complete, so the
+consumption of the argument need not be either. However, because
+exceptions can be caught, a program can observe a state where a value
+``v`` has been passed to a linear function ``f`` but the call ``f v``
+has exited (with an exception) without consuming ``v``. So while, the
+guarantee provided by linear functions holds for converging
+computations, we must weaken it in case of divergence:
+
+- Attempting to consume exactly once ``f v``, when ``f`` is a linear
+  function, will consume ``v`` exactly once if the consumption of ``f
+  v`` converges, and *at most once* if it diverges.
+
+Where "consuming at most once" is defined by induction, like
+"consuming exactly once", but every sub-consumption is optional.
+
+In the paper, we gave a simplified specification of a linear ``IO``
+monad (called ``IOL``) which ignored the issue of exception for the
+sake of simplicity. Can we, still, write a resource-safe ``RIO`` monad
+with linear types despite the added difficulty of exceptions? Yes, as
+this section will show.
```

Comment:
Opening a subthread here to answer @simonmar 's concerns.

>Most resources are allocated with a bracket pattern (`bracket allocate deallocate $ \resource -> do ...`) or using a `withFoo combinator`. These give clear scoping and exception-safety to resource allocation, and linear types don't offer any safety improvements here - as I understand it the bracket pattern will remain the safest way to manage explicit resources in cases where it can be used.

The goal of this section, currently §2.14 (Non-termination, exceptions & catch), and of the [corresponding library](https://github.com/tweag/linear-base/blob/master/src/System/IO/Resource.hs) is to demonstrate that, actually, it's quite possible to build a linear abstraction which will be safer than brackets.

> The ones that spring to mind are where resources are allocated in C++ and released in Haskell or allocated in Haskell and then released in a different Haskell thread after receiving a message from C++.

It sounds like a similar use-case as [this blog post](https://www.tweag.io/posts/2017-11-29-linear-jvm.html).

> our experience has been that safety in these cases has only been an issue where exceptions are concerned - and linear types doesn't help with that

Does section §2.14 convince you that, in fact, linear types can help you build abstractions where the program logic doesn't have to deal with exceptions, while ensuring proper release?